x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -277,6 +290,88 @@ public void removeFromCache(final CacheKey cacheKey) {
+                    runIfShardMarkedAsEvictedInCache(shardEviction, () -> {
+                        assert shardsEvictionLock.isHeldByCurrentThread(shardEviction);
+                        final Map<CacheKey, CacheFile> cacheFilesToEvict = new HashMap<>();
+                        cache.forEach((cacheKey, cacheFile) -> {
+                            if (shardEviction.matches(cacheKey)) {
+                                cacheFilesToEvict.put(cacheKey, cacheFile);
+                            }
+                        });
+                        if (cacheFilesToEvict.isEmpty() == false) {
+                            cacheFilesToEvict.forEach(cache::invalidate);
+                        }
+                    });
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.warn(
+                        () -> new ParameterizedMessage("failed to evict cache files associated with evicted shard {}", shardEviction),
+                        e
+                    );
+                }
+            });
+        }
+    }
+
+    /**
+     * Allows to run the specified {@link Runnable} if the shard represented by the triplet ({@link SnapshotId}, {@link IndexId},
+     * {@link SnapshotId}) is still marked as evicted at the time this method is executed. The @link Runnable} will be executed
+     * while the current thread is holding the lock associated to the shard.
+     *
+     * @param snapshotId the snapshot the evicted searchable snapshots shard belongs to
+     * @param indexId    the index in the snapshot the evicted searchable snapshots shard belongs to
+     * @param shardId    the searchable snapshots shard id
+     * @param runnable   a runnable to execute
+     */
+    public void runIfShardMarkedAsEvictedInCache(SnapshotId snapshotId, IndexId indexId, ShardId shardId, Runnable runnable) {
+        runIfShardMarkedAsEvictedInCache(new ShardEviction(snapshotId, indexId, shardId), runnable);
+    }
+
+    /**
+     * Allows to run the specified {@link Runnable} if the shard represented by {@link ShardEviction} is still marked as evicted at the time
+     * this method is executed. The @link Runnable} will be executed while the current thread is holding the lock associated to the shard.
+     *
+     * @param shardEviction a {@link ShardEviction} representing the shard marked as evicted
+     * @param runnable      a runnable to execute
+     */
+    private void runIfShardMarkedAsEvictedInCache(ShardEviction shardEviction, Runnable runnable) {
+        try (Releasable ignored = shardsEvictionLock.acquire(shardEviction)) {
+            boolean success = false;
+            try {
+                if (evictedShards.remove(shardEviction)) {
+                    runnable.run();
+                }
+                success = true;
+            } finally {
+                if (success == false) {
+                    final boolean added = evictedShards.add(shardEviction);
```

Comment:
I think you are right, I pushed 393deaf6621cc2fd7650fc946f1344a0f3bd53d9 to remove the usage of the lock here (but such locks are still necessary to avoid cache files eviction while starting the directory). Let me know what you think.

Reply:
I think the locks could be removed with some restructure, but we can tackle that in a follow-up, it is not really important.

However, if we add to `evictedShards` here while running the job on the thread, I think we risk it leaking since nothing will remove it later? I think we should `assert success` here. AFAICS, there is no way this should fail unless there is a bug somewhere.

Reply:
I will be happy to know more about how you think it could be restructured. For now I'm taking this point and will come back to you in a short future for tackling this.

I agree there's a risk of leaking a `ShardEviction` around but I think it is very unlikely to happen. One idea would be to hook into the cache sync task to clean up any left overs `ShardEviction` there. If that's OK we can address this too as a follow up with the previous point. For now I added the success assertion.

Reply:
I've opened #67160 which should improve the situation.